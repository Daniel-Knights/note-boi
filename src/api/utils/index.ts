export * from './decorators';
export * from './error';
export * from './http';
