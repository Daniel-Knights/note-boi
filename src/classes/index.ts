export * from './appError';
export * from './debounceQueue';
export * from './encryptor';
export * from './fetchBuilder';
export * from './keyStore';
export * from './storage';
export * from './unsyncedNotesManager';
