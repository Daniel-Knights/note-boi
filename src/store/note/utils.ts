import { Note } from '../../classes';
import { isEmptyNote } from '../../utils';

import { deleteNote } from './operations';
import { noteState } from './state';

/** Sorts notes in descending order by timestamp. */
export const sortNotesFn = (a: Note, b: Note): number => b.timestamp - a.timestamp;

/** Sorts {@link noteState.notes} in descending order by timestamp. */
export const sortStateNotes = (): Note[] => noteState.notes.sort(sortNotesFn);

/** Finds note index within {@link noteState.notes}. */
export function findNoteIndex(uuid?: string): number {
  return noteState.notes.findIndex((nt) => nt.uuid === uuid);
}

/** Finds note within {@link noteState.notes}. */
export function findNote(uuid?: string): Note | undefined {
  return noteState.notes.find((nt) => nt.uuid === uuid);
}

/**
 * Returns true if note is either {@link noteState.selectedNote}
 * or within {@link noteState.extraSelectedNotes}.
 */
export function isSelectedNote(note: Note): boolean {
  return (
    note.uuid === noteState.selectedNote.uuid ||
    noteState.extraSelectedNotes.some((nt) => nt?.uuid === note.uuid)
  );
}

/** Deletes {@link noteState.selectedNote} when note is empty. */
export function clearEmptyNote(): void {
  const isValidClear = noteState.notes.length > 1;
  if (!isValidClear) return;

  const foundNote = findNote(noteState.selectedNote.uuid);
  if (!foundNote) return;

  if (isEmptyNote(foundNote)) {
    deleteNote(noteState.selectedNote.uuid);
  }
}
