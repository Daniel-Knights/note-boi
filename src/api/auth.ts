import {
  AppError,
  Encryptor,
  ERROR_CODE,
  ErrorConfig,
  FetchBuilder,
  KeyStore,
  Storage,
} from '../classes';
import { noteState } from '../store/note';
import { resetAppError, syncState } from '../store/sync';
import { isEmptyNote, tauriEmit, tauriInvoke } from '../utils';

import { updateLocalNoteStateFromDiff } from './notes';
import {
  parseErrorRes,
  resIsOk,
  route,
  throwEncryptorError,
  throwFetchError,
} from './utils';

export function clientSideLogout(): Promise<void> {
  tauriInvoke('delete_access_token', { username: syncState.username });

  syncState.username = '';
  syncState.isLoggedIn = false;

  Storage.remove('USERNAME');
  KeyStore.reset();

  return tauriEmit('auth', { is_logged_in: false });
}

// Login
export const login = route(async () => {
  const errorConfig = {
    code: ERROR_CODE.LOGIN,
    retry: { fn: login },
    display: {
      form: true,
      sync: true,
    },
  } satisfies Omit<ErrorConfig<typeof login>, 'message'>;

  const passwordKey = await Encryptor.generatePasswordKey(syncState.password);

  const encryptedNotes = await Encryptor.encryptNotes(
    noteState.notes.filter((nt) => !isEmptyNote(nt)),
    passwordKey
  ).catch((err) => throwEncryptorError(errorConfig, err));
  if (!encryptedNotes) return;

  encryptedNotes.forEach((nt) => {
    syncState.encryptedNotesCache.set(nt.id, nt);
  });

  const res = await new FetchBuilder('/auth/login')
    .method('POST')
    .body({
      username: syncState.username,
      password: syncState.password,
      notes: encryptedNotes,
      deleted_notes: syncState.unsyncedNotes.deleted,
    })
    .fetch(syncState.username)
    .catch((err) => throwFetchError(errorConfig, err));
  if (!res) return;

  if (resIsOk(res)) {
    syncState.password = '';
    syncState.isLoggedIn = true;

    Storage.set('USERNAME', syncState.username);

    resetAppError();
    tauriEmit('auth', { is_logged_in: true });

    await KeyStore.storeKey(passwordKey);

    const decryptedNotes = await Promise.all([
      Encryptor.decryptNotes(res.data.note_diff.added, passwordKey),
      Encryptor.decryptNotes(res.data.note_diff.edited, passwordKey),
    ]).catch((err) => throwEncryptorError(errorConfig, err));
    if (!decryptedNotes) return;

    await updateLocalNoteStateFromDiff({
      added: decryptedNotes[0],
      edited: decryptedNotes[1],
      deleted: res.data.note_diff.deleted,
    });
  } else {
    throw new AppError({
      ...errorConfig,
      message: parseErrorRes(res),
    });
  }
});

// Signup
export const signup = route(async () => {
  const errorConfig = {
    code: ERROR_CODE.SIGNUP,
    retry: { fn: signup },
    display: {
      form: true,
      sync: true,
    },
  } satisfies Omit<ErrorConfig<typeof signup>, 'message'>;

  const passwordKey = await Encryptor.generatePasswordKey(syncState.password);

  await KeyStore.storeKey(passwordKey);

  const encryptedNotes = await Encryptor.encryptNotes(
    noteState.notes.filter((nt) => !isEmptyNote(nt)),
    passwordKey
  ).catch((err) => throwEncryptorError(errorConfig, err));
  if (!encryptedNotes) return;

  const res = await new FetchBuilder('/auth/signup')
    .method('POST')
    .body({
      username: syncState.username,
      password: syncState.password,
      notes: encryptedNotes,
    })
    .fetch(syncState.username)
    .catch((err) => throwFetchError(errorConfig, err));
  if (!res) return;

  if (resIsOk(res)) {
    resetAppError();
    tauriEmit('auth', { is_logged_in: true });

    syncState.password = '';
    syncState.isLoggedIn = true;
    syncState.unsyncedNotes.clear();

    Storage.set('USERNAME', syncState.username);
  } else {
    throw new AppError({
      ...errorConfig,
      message: parseErrorRes(res),
    });
  }
});

// Logout
export const logout = route(async () => {
  const errorConfig = {
    code: ERROR_CODE.LOGOUT,
    retry: { fn: logout },
    display: { form: true },
  } satisfies Omit<ErrorConfig<typeof logout>, 'message'>;

  const accessToken = await tauriInvoke('get_access_token', {
    username: syncState.username,
  });

  const fetchPromise = new FetchBuilder('/auth/logout')
    .method('POST')
    .withAuth(syncState.username, accessToken)
    .fetch()
    .catch((err) => throwFetchError(errorConfig, err));

  const [res] = await Promise.all([fetchPromise, clientSideLogout()]);
  if (!res) return;

  if (resIsOk(res)) {
    resetAppError();
  } else {
    // No use bringing this to the user's attention
    throw new AppError({
      code: ERROR_CODE.LOGOUT,
      message: parseErrorRes(res),
    });
  }
});
