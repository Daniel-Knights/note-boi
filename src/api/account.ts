import * as dialog from '@tauri-apps/plugin-dialog';

import {
  AppError,
  Encryptor,
  ERROR_CODE,
  ErrorConfig,
  FetchBuilder,
  KeyStore,
} from '../classes';
import { noteState } from '../store/note';
import { resetAppError, syncState } from '../store/sync';
import { tauriInvoke } from '../utils';

import { clientSideLogout } from './auth';
import {
  parseErrorRes,
  resIsOk,
  route,
  throwEncryptorError,
  throwFetchError,
} from './utils';

export const changePassword = route(async (): Promise<void> => {
  const errorConfig = {
    code: ERROR_CODE.CHANGE_PASSWORD,
    retry: { fn: changePassword },
    display: {
      form: true,
      sync: true,
    },
  } satisfies Omit<ErrorConfig<typeof changePassword>, 'message'>;

  const newKey = await Encryptor.generatePasswordKey(syncState.newPassword);

  await KeyStore.storeKey(newKey);

  const [accessToken, encryptedNotes] = await Promise.all([
    tauriInvoke('get_access_token', {
      username: syncState.username,
    }),
    Encryptor.encryptNotes(noteState.notes, newKey).catch((err) =>
      throwEncryptorError(errorConfig, err)
    ),
  ]);
  if (!accessToken || !encryptedNotes) return;

  const res = await new FetchBuilder('/account/change-password')
    .method('PUT')
    .withAuth(syncState.username, accessToken)
    .body({
      current_password: syncState.password,
      new_password: syncState.newPassword,
      notes: encryptedNotes,
    })
    .fetch(syncState.username)
    .catch((err) => throwFetchError(errorConfig, err));
  if (!res) return;

  if (resIsOk(res)) {
    resetAppError();

    syncState.password = '';
    syncState.newPassword = '';
  } else {
    throw new AppError({
      ...errorConfig,
      message: parseErrorRes(res),
    });
  }
});

export const deleteAccount = route(async (): Promise<void> => {
  const askRes = await dialog.ask('Are you sure?', {
    title: 'Delete account',
    kind: 'warning',
  });
  if (!askRes) return;

  const errorConfig = {
    code: ERROR_CODE.DELETE_ACCOUNT,
    retry: { fn: deleteAccount },
    display: {
      sync: true,
    },
  } satisfies Omit<ErrorConfig<typeof deleteAccount>, 'message'>;

  const accessToken = await tauriInvoke('get_access_token', {
    username: syncState.username,
  });

  const res = await new FetchBuilder('/account/delete')
    .method('DELETE')
    .withAuth(syncState.username, accessToken)
    .fetch()
    .catch((err) => throwFetchError(errorConfig, err));
  if (!res) return;

  if (resIsOk(res)) {
    await clientSideLogout();
    resetAppError();
    syncState.unsyncedNotes.clear(true);
  } else {
    throw new AppError({
      ...errorConfig,
      message: parseErrorRes(res),
    });
  }
});
