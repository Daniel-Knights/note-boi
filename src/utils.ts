import { event, invoke } from '@tauri-apps/api';
import { EventCallback, UnlistenFn } from '@tauri-apps/api/event';

import { TauriCommand } from './constant';
import type { Note } from './store/note';

/** `process.env.NODE_ENV === 'development'`. */
export function isDev(): boolean {
  return process.env.NODE_ENV === 'development';
}

/** Formats Unix time to date-time. */
export function unixToDateTime(unixTime: number): string {
  return Intl.DateTimeFormat([], {
    dateStyle: 'medium',
    timeStyle: isDev() ? 'long' : 'short',
  }).format(unixTime);
}

/** Gets key value from `localStorage` and parses it as JSON. */
export function localStorageParse(key: string): ReturnType<typeof JSON.parse> {
  return JSON.parse(localStorage.getItem(key) || '{}');
}

/** Returns `true` if string consists of only whitespace characters or is empty. */
export function isWhitespaceOnly(text?: string): boolean {
  return text?.trim() === '';
}

/** Returns true if note content is empty or note is `undefined`. */
export function isEmptyNote(note?: Note): boolean {
  if (!note) return true;

  return isWhitespaceOnly(note.content.title) && isWhitespaceOnly(note.content.body);
}

/** Calls {@link event.emit}, with stronger typing for `id`. */
export function tauriEmit<T>(id: 'login' | 'logout', payload?: T): void {
  event.emit(id, payload);
}

/** Calls {@link event.listen}, with stronger typing for `id`. */
export function tauriListen(
  id:
    | 'reload'
    | 'new-note'
    | 'delete-note'
    | 'push-notes'
    | 'login'
    | 'logout'
    | 'signup'
    | `tauri://${string}`,
  cb: EventCallback<unknown>
): Promise<UnlistenFn> {
  return event.listen(id, cb);
}

/** Calls {@link invoke}, with stronger typing for `cmd`. */
export function tauriInvoke<T>(
  cmd: TauriCommand,
  args?: Parameters<typeof invoke>[1]
): Promise<T> {
  return invoke(cmd, args);
}
