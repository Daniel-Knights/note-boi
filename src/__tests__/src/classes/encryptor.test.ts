import { Encryptor, KeyStore } from '../../../classes';
import { mockDb } from '../../api';
import localNotes from '../../notes.json';
import { isNote } from '../../utils';

describe('Encryptor', () => {
  describe('setPasswordKey', () => {
    it('Generates and stores a CryptoKey from the given password', async () => {
      const password = '1';

      await Encryptor.setPasswordKey(password);

      const passwordKey = await KeyStore.getKey();

      assert.isDefined(passwordKey);
    });
  });

  describe('encryptNotes', () => {
    it('Sets password key if password is given', async () => {
      const setPasswordKeySpy = vi.spyOn(Encryptor, 'setPasswordKey');

      await Encryptor.encryptNotes([], '1');

      expect(setPasswordKeySpy).toHaveBeenCalled();
    });

    it("Doesn't set password key if no password is given", async () => {
      const setPasswordKeySpy = vi.spyOn(Encryptor, 'setPasswordKey');

      await Encryptor.encryptNotes([]);

      expect(setPasswordKeySpy).not.toHaveBeenCalled();
    });

    it('Encrypts notes', async () => {
      const encryptedNotes = await Encryptor.encryptNotes(localNotes, '1');

      assert.lengthOf(encryptedNotes, localNotes.length);

      encryptedNotes.forEach((note) => {
        assert.isString(note.content);
      });
    });
  });

  describe('decryptNotes', () => {
    it('Sets password key if password is given', async () => {
      const setPasswordKeySpy = vi.spyOn(Encryptor, 'setPasswordKey');

      await Encryptor.decryptNotes([], '1');

      expect(setPasswordKeySpy).toHaveBeenCalled();
    });

    it("Doesn't set password key if no password is given", async () => {
      const setPasswordKeySpy = vi.spyOn(Encryptor, 'setPasswordKey');

      await Encryptor.decryptNotes([]);

      expect(setPasswordKeySpy).not.toHaveBeenCalled();
    });

    it('Decrypts notes', async () => {
      const decryptedNotes = await Encryptor.decryptNotes(mockDb.encryptedNotes, '1');

      assert.lengthOf(decryptedNotes, mockDb.encryptedNotes.length);

      decryptedNotes.forEach((nt) => {
        assert.isTrue(isNote(nt));
      });
    });

    it('Handles unencrypted notes', async () => {
      const decryptedNotes = await Encryptor.decryptNotes(localNotes);

      assert.lengthOf(decryptedNotes, localNotes.length);

      decryptedNotes.forEach((nt) => {
        assert.isTrue(isNote(nt));
      });
    });
  });
});
