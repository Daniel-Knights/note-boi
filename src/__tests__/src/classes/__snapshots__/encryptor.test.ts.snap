// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Encryptor > decryptNotes > Decrypts notes > KeyStore.getKey() 1`] = `CryptoKey {}`;

exports[`Encryptor > decryptNotes > Decrypts notes > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Encryptor > decryptNotes > Decrypts notes > Storage.get('THEME') 1`] = `null`;

exports[`Encryptor > decryptNotes > Decrypts notes > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Encryptor > decryptNotes > Decrypts notes > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Encryptor > decryptNotes > Decrypts notes > Storage.get('USERNAME') 1`] = `null`;

exports[`Encryptor > decryptNotes > Decrypts notes > allCalls 1`] = `[]`;

exports[`Encryptor > decryptNotes > Decrypts notes > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`Encryptor > decryptNotes > Decrypts notes > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`Encryptor > decryptNotes > Decrypts notes > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`Encryptor > decryptNotes > Decrypts notes > openedPopup 1`] = `undefined`;

exports[`Encryptor > decryptNotes > Decrypts notes > selectedTheme 1`] = `"System"`;

exports[`Encryptor > decryptNotes > Decrypts notes > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoading": false,
  "isLoggedIn": false,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`Encryptor > decryptNotes > Decrypts notes > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`Encryptor > decryptNotes > Doesn't set password key if no password is given > KeyStore.getKey() 1`] = `undefined`;

exports[`Encryptor > decryptNotes > Doesn't set password key if no password is given > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Encryptor > decryptNotes > Doesn't set password key if no password is given > Storage.get('THEME') 1`] = `null`;

exports[`Encryptor > decryptNotes > Doesn't set password key if no password is given > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Encryptor > decryptNotes > Doesn't set password key if no password is given > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Encryptor > decryptNotes > Doesn't set password key if no password is given > Storage.get('USERNAME') 1`] = `null`;

exports[`Encryptor > decryptNotes > Doesn't set password key if no password is given > allCalls 1`] = `[]`;

exports[`Encryptor > decryptNotes > Doesn't set password key if no password is given > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`Encryptor > decryptNotes > Doesn't set password key if no password is given > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`Encryptor > decryptNotes > Doesn't set password key if no password is given > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`Encryptor > decryptNotes > Doesn't set password key if no password is given > openedPopup 1`] = `undefined`;

exports[`Encryptor > decryptNotes > Doesn't set password key if no password is given > selectedTheme 1`] = `"System"`;

exports[`Encryptor > decryptNotes > Doesn't set password key if no password is given > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoading": false,
  "isLoggedIn": false,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`Encryptor > decryptNotes > Doesn't set password key if no password is given > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`Encryptor > decryptNotes > Handles unencrypted notes > KeyStore.getKey() 1`] = `undefined`;

exports[`Encryptor > decryptNotes > Handles unencrypted notes > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Encryptor > decryptNotes > Handles unencrypted notes > Storage.get('THEME') 1`] = `null`;

exports[`Encryptor > decryptNotes > Handles unencrypted notes > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Encryptor > decryptNotes > Handles unencrypted notes > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Encryptor > decryptNotes > Handles unencrypted notes > Storage.get('USERNAME') 1`] = `null`;

exports[`Encryptor > decryptNotes > Handles unencrypted notes > allCalls 1`] = `[]`;

exports[`Encryptor > decryptNotes > Handles unencrypted notes > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`Encryptor > decryptNotes > Handles unencrypted notes > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`Encryptor > decryptNotes > Handles unencrypted notes > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`Encryptor > decryptNotes > Handles unencrypted notes > openedPopup 1`] = `undefined`;

exports[`Encryptor > decryptNotes > Handles unencrypted notes > selectedTheme 1`] = `"System"`;

exports[`Encryptor > decryptNotes > Handles unencrypted notes > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoading": false,
  "isLoggedIn": false,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`Encryptor > decryptNotes > Handles unencrypted notes > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`Encryptor > decryptNotes > Sets password key if password is given > KeyStore.getKey() 1`] = `CryptoKey {}`;

exports[`Encryptor > decryptNotes > Sets password key if password is given > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Encryptor > decryptNotes > Sets password key if password is given > Storage.get('THEME') 1`] = `null`;

exports[`Encryptor > decryptNotes > Sets password key if password is given > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Encryptor > decryptNotes > Sets password key if password is given > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Encryptor > decryptNotes > Sets password key if password is given > Storage.get('USERNAME') 1`] = `null`;

exports[`Encryptor > decryptNotes > Sets password key if password is given > allCalls 1`] = `[]`;

exports[`Encryptor > decryptNotes > Sets password key if password is given > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`Encryptor > decryptNotes > Sets password key if password is given > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`Encryptor > decryptNotes > Sets password key if password is given > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`Encryptor > decryptNotes > Sets password key if password is given > openedPopup 1`] = `undefined`;

exports[`Encryptor > decryptNotes > Sets password key if password is given > selectedTheme 1`] = `"System"`;

exports[`Encryptor > decryptNotes > Sets password key if password is given > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoading": false,
  "isLoggedIn": false,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`Encryptor > decryptNotes > Sets password key if password is given > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`Encryptor > encryptNotes > Doesn't set password key if no password is given > KeyStore.getKey() 1`] = `undefined`;

exports[`Encryptor > encryptNotes > Doesn't set password key if no password is given > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Encryptor > encryptNotes > Doesn't set password key if no password is given > Storage.get('THEME') 1`] = `null`;

exports[`Encryptor > encryptNotes > Doesn't set password key if no password is given > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Encryptor > encryptNotes > Doesn't set password key if no password is given > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Encryptor > encryptNotes > Doesn't set password key if no password is given > Storage.get('USERNAME') 1`] = `null`;

exports[`Encryptor > encryptNotes > Doesn't set password key if no password is given > allCalls 1`] = `[]`;

exports[`Encryptor > encryptNotes > Doesn't set password key if no password is given > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`Encryptor > encryptNotes > Doesn't set password key if no password is given > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`Encryptor > encryptNotes > Doesn't set password key if no password is given > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`Encryptor > encryptNotes > Doesn't set password key if no password is given > openedPopup 1`] = `undefined`;

exports[`Encryptor > encryptNotes > Doesn't set password key if no password is given > selectedTheme 1`] = `"System"`;

exports[`Encryptor > encryptNotes > Doesn't set password key if no password is given > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoading": false,
  "isLoggedIn": false,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`Encryptor > encryptNotes > Doesn't set password key if no password is given > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`Encryptor > encryptNotes > Encrypts notes > KeyStore.getKey() 1`] = `CryptoKey {}`;

exports[`Encryptor > encryptNotes > Encrypts notes > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Encryptor > encryptNotes > Encrypts notes > Storage.get('THEME') 1`] = `null`;

exports[`Encryptor > encryptNotes > Encrypts notes > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Encryptor > encryptNotes > Encrypts notes > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Encryptor > encryptNotes > Encrypts notes > Storage.get('USERNAME') 1`] = `null`;

exports[`Encryptor > encryptNotes > Encrypts notes > allCalls 1`] = `[]`;

exports[`Encryptor > encryptNotes > Encrypts notes > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`Encryptor > encryptNotes > Encrypts notes > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`Encryptor > encryptNotes > Encrypts notes > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`Encryptor > encryptNotes > Encrypts notes > openedPopup 1`] = `undefined`;

exports[`Encryptor > encryptNotes > Encrypts notes > selectedTheme 1`] = `"System"`;

exports[`Encryptor > encryptNotes > Encrypts notes > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoading": false,
  "isLoggedIn": false,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`Encryptor > encryptNotes > Encrypts notes > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`Encryptor > encryptNotes > Sets password key if password is given > KeyStore.getKey() 1`] = `CryptoKey {}`;

exports[`Encryptor > encryptNotes > Sets password key if password is given > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Encryptor > encryptNotes > Sets password key if password is given > Storage.get('THEME') 1`] = `null`;

exports[`Encryptor > encryptNotes > Sets password key if password is given > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Encryptor > encryptNotes > Sets password key if password is given > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Encryptor > encryptNotes > Sets password key if password is given > Storage.get('USERNAME') 1`] = `null`;

exports[`Encryptor > encryptNotes > Sets password key if password is given > allCalls 1`] = `[]`;

exports[`Encryptor > encryptNotes > Sets password key if password is given > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`Encryptor > encryptNotes > Sets password key if password is given > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`Encryptor > encryptNotes > Sets password key if password is given > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`Encryptor > encryptNotes > Sets password key if password is given > openedPopup 1`] = `undefined`;

exports[`Encryptor > encryptNotes > Sets password key if password is given > selectedTheme 1`] = `"System"`;

exports[`Encryptor > encryptNotes > Sets password key if password is given > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoading": false,
  "isLoggedIn": false,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`Encryptor > encryptNotes > Sets password key if password is given > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`Encryptor > setPasswordKey > Generates and stores a CryptoKey from the given password > KeyStore.getKey() 1`] = `CryptoKey {}`;

exports[`Encryptor > setPasswordKey > Generates and stores a CryptoKey from the given password > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Encryptor > setPasswordKey > Generates and stores a CryptoKey from the given password > Storage.get('THEME') 1`] = `null`;

exports[`Encryptor > setPasswordKey > Generates and stores a CryptoKey from the given password > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Encryptor > setPasswordKey > Generates and stores a CryptoKey from the given password > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Encryptor > setPasswordKey > Generates and stores a CryptoKey from the given password > Storage.get('USERNAME') 1`] = `null`;

exports[`Encryptor > setPasswordKey > Generates and stores a CryptoKey from the given password > allCalls 1`] = `[]`;

exports[`Encryptor > setPasswordKey > Generates and stores a CryptoKey from the given password > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`Encryptor > setPasswordKey > Generates and stores a CryptoKey from the given password > mockDb 1`] = `
{
  "encryptedNotes": 10,
  "users": {
    "d": "1",
  },
}
`;

exports[`Encryptor > setPasswordKey > Generates and stores a CryptoKey from the given password > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "id": "id",
    "timestamp": 0,
  },
}
`;

exports[`Encryptor > setPasswordKey > Generates and stores a CryptoKey from the given password > openedPopup 1`] = `undefined`;

exports[`Encryptor > setPasswordKey > Generates and stores a CryptoKey from the given password > selectedTheme 1`] = `"System"`;

exports[`Encryptor > setPasswordKey > Generates and stores a CryptoKey from the given password > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "isLoading": false,
  "isLoggedIn": false,
  "newPassword": "",
  "password": "",
  "unsyncedNoteIds": {
    "add": [Function],
    "clear": [Function],
    "deleted": Set {},
    "edited": Set {},
    "new": "",
    "size": 0,
  },
  "username": "",
}
`;

exports[`Encryptor > setPasswordKey > Generates and stores a CryptoKey from the given password > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;
